factor/sms/classes/factor.php

```diff
@@ -0,0 +1,241 @@
+            'userid' => $user->id,
+            'factor' => $this->name,
+            'createdfromip' => $user->lastip,
+            'timecreated' => time(),
+            'revoked' => 0,
+        );
+        $record['id'] = $DB->insert_record('tool_mfa', $record, true);
+        return [(object) $record];
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_input() {
+        return true;
+    }
+
+    /**
+     * SMS Factor implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function has_setup() {
+        return true;
+    }
+
+    /**
+     * Generates and emails the code for login to the user, stores codes in DB.
+     *
+     * @return void
+     */
+    private function generate_and_sms_code() {
+        global $DB, $USER;
+
+        // Get instance that isnt parent SMS type (empty label check).
+        // This check must exclude the main singleton record, with the label as the email.
+        // It must only grab the record with the user agent as the label.
+        $sql = 'SELECT *
+                  FROM {tool_mfa}
+                 WHERE userid = ?
+                   AND factor = ?
+               AND label IS NOT NULL';
+
+        $record = $DB->get_record_sql($sql, array($USER->id, 'sms'));
+        $duration = get_config('factor_sms', 'duration');
+        $newcode = random_int(100000, 999999);
+
+        $number = $USER->phone2;
+
+        if (empty($record)) {
+            // No code active, generate new code.
+            $instanceid = $DB->insert_record('tool_mfa', array(
+                'userid' => $USER->id,
+                'factor' => 'sms',
+                'secret' => $newcode,
+                'label' => $number,
+                'timecreated' => time(),
+                'createdfromip' => $USER->lastip,
+                'timemodified' => time(),
+                'lastverified' => time(),
+                'revoked' => 0,
+            ), true);
+            $this->sms_verification_code($instanceid);
+
+        } else if ($record->timecreated + $duration < time()) {
+            // Old code found. Keep id, update fields.
+            $DB->update_record('tool_mfa', array(
+                'id' => $record->id,
+                'secret' => $newcode,
+                'label' => $number,
+                'timecreated' => time(),
+                'createdfromip' => $USER->lastip,
+                'timemodified' => time(),
+                'lastverified' => time(),
+                'revoked' => 0,
+            ));
+            $instanceid = $record->id;
+            $this->sms_verification_code($instanceid);
+        }
+    }
+
+    private function sms_verification_code($instanceid) {
+        global $DB;
+
+        // Here we should get the information, then construct the message.
+        $instance = $DB->get_record('tool_mfa', ['id' => $instanceid]);
+        $user = \core_user::get_user($instance->userid);
+        $phonenumber = $user->phone2;
```

Comment:
So I'm quite conflicted about using this profile field vs storing the number in the factor table like totp secret. Pros and cons:

Profile:

pros: it makes sense to the user
cons: zero validation, it can change at any time by the user. It could be the word 'frog' or have a # and extension numbers etc

Factor:

pro: full contol over validation
cons: different UX

So I'm open to all discussion but I'm leaning towards some sort of hybrid:

1) the user can supply the phone (eg on the registration page)
2) the MFA factor never trusts this, it must confirm it before using it the same as totp
3) if the factor does not have a stored number, but it detects that the profile does have one then will send a text to confirm
4) user confirms like totp, at this point the number is cloned into the factor table and we display it as something like 'confirmed number'
5) bonus points we can register event for when the profile fields have changed and compare to the confirmed number, if it is different then we can register a second number and probably leave the previous one in place to be manually removed in the mfa preferences

I think as bonus points the setup_factor_form_definition($mform) should exist and show a form which directly edits the mobile phone profile field, but which also adds some validation to it. The validation doesn't need to be perfect as we will be sending an sms to it very soon after. We should be fairly liberal with accepting ( and ) - etc and removing them. The challenge is we can't do this validation directly on the profile unless we make a sibling custom profile field type which is a whole new plugin which sucks.

This looks promising:

https://www.oreilly.com/library/view/regular-expressions-cookbook/9781449327453/ch04s03.html